Errors on nodes

MichaelMauderer opened the pull request:
### Pull Request Description
Adds a graph editor API to indicate that nodes have an error.

![Screenshot from 2020-12-22 16-03-09](https://user-images.githubusercontent.com/1428930/102907558-83067400-4476-11eb-8eb4-2f92a5bf2d6d.png)
![Screenshot from 2020-12-22 16-08-47](https://user-images.githubusercontent.com/1428930/102907562-839f0a80-4476-11eb-8be8-4b37099e98fe.png)

![Screenshot from 2020-12-22 16-02-54](https://user-images.githubusercontent.com/1428930/102907552-81d54700-4476-11eb-861d-a83fda31df72.png)


### Important Notes
Provisional API that might need updating based on the language server specification of errors.

Currently, can only be seen in the debug scene as the language server API does not exist yet. 

The error is shown above the node to not overlap the visualizations.

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guide.
- [x] All code has automatic tests where possible.
- [x] All code has been profiled where possible.
- [x] All code has been manually tested in the IDE.
- [x] All code has been manually tested in the "debug/interface" scene.
- [x] All code has been manually tested by the PR owner against our [test scenarios](https://docs.google.com/spreadsheets/d/1RatJDM_f9_3bvYhl3Bpq2d8SyKgtVdrV1RkGxPU17c8/edit?ts=5faa7049#gid=0).
- [ ] All code has been manually tested by at least one reviewer against our [test scenarios](https://docs.google.com/spreadsheets/d/1RatJDM_f9_3bvYhl3Bpq2d8SyKgtVdrV1RkGxPU17c8/edit?ts=5faa7049#gid=0).

wdanilo commented on src/rust/ide/view/graph-editor/src/component/node.rs:
```diff
@@ -137,6 +150,33 @@ pub mod drag_area {
     }
 }
 
+pub mod error_shape {
```
this should not be in the "shape" section, as this is not node shape anymore

wdanilo commented on src/rust/ide/view/graph-editor/src/component/node.rs:
```diff
@@ -137,6 +150,33 @@ pub mod drag_area {
+
+            let repeat_x      = ERROR_PATTERN_REPEAT_X.px();
+            let repeat_y      = ERROR_PATTERN_REPEAT_Y.px();
+            let repeat        =  Var::<Vector2<Pixels>>::from((repeat_x,repeat_y));
+            let error_width   =  ERROR_PATTERN_STRIPE_WIDTH.px();
```
alignment

wdanilo commented on src/rust/ide/view/graph-editor/src/component/node.rs:
```diff
@@ -137,6 +150,33 @@ pub mod drag_area {
+            let repeat        =  Var::<Vector2<Pixels>>::from((repeat_x,repeat_y));
+            let error_width   =  ERROR_PATTERN_STRIPE_WIDTH.px();
+            let stripe_red    = Rect((error_width, INFINITE.px()));
+            let pattern       = stripe_red.repeat(repeat).rotate(45.0.radians());
```
make it constant pls

wdanilo commented on src/rust/ide/view/graph-editor/src/component/node.rs:
```diff
@@ -137,6 +150,33 @@ pub mod drag_area {
+pub mod error_shape {
+    use super::*;
+
+    ensogl::define_shape_system! {
```
does it implement wider stripes on zoom out?

wdanilo commented on src/rust/ide/view/graph-editor/src/component/node.rs:
```diff
@@ -273,15 +314,17 @@ impl Deref for Node {
+    pub logger          : Logger,
+    pub main_area       : component::ShapeView<shape::Shape>,
+    pub drag_area       : component::ShapeView<drag_area::Shape>,
+    pub error_indicator : component::ShapeView<error_shape::Shape>,
```
why not `error_area`? I believe it is an area, looking at the screenshots, and then it would have consistent name with other areas.

MichaelMauderer replied:
I find `indicator` more apt. It has no area properties, instead, it something that can visually indicate that there is an error. It's only states are on/off (with animations) and we don't care about area-based interactions with it.

wdanilo commented on src/rust/ide/view/graph-editor/src/component/node.rs:
```diff
@@ -292,17 +335,28 @@ impl NodeModel {
-        let drag_logger = Logger::sub(&logger,"drag_area");
-        let main_area   = component::ShapeView::<shape::Shape>::new(&main_logger,scene);
-        let drag_area   = component::ShapeView::<drag_area::Shape>::new(&drag_logger,scene);
+        let main_logger             = Logger::sub(&logger,"main_area");
+        let drag_logger             = Logger::sub(&logger,"drag_area");
+        let error_indicator_logger  = Logger::sub(&logger,"drag_area");
```
these namesare inconsistent

wdanilo commented on src/rust/ide/view/graph-editor/src/component/node.rs:
```diff
@@ -292,17 +335,28 @@ impl NodeModel {
+        let drag_logger             = Logger::sub(&logger,"drag_area");
+        let error_indicator_logger  = Logger::sub(&logger,"drag_area");
+
+        let error_indicator = component::ShapeView::<error_shape::Shape>::new(&error_indicator_logger,scene);
```
long line. Please mmake sure that all new lines < 100 chars

wdanilo commented on src/rust/ide/view/graph-editor/src/component/node.rs:
```diff
@@ -292,17 +335,28 @@ impl NodeModel {
+        display_object.add_child(&error_text);
+
+        error_text.set_default_color.emit(color::Rgba::red());
+        let text_size:TextSize    = TEXT_SIZE.into();
```
alignment + you can erge this line with the next one

wdanilo commented on src/rust/ide/view/graph-editor/src/component/node.rs:
```diff
@@ -325,7 +379,8 @@ impl NodeModel {
-        Self {app,display_object,logger,main_area,drag_area,output,input,visualization,action_bar}
+        Self {app,display_object,logger,main_area,drag_area,output,input,visualization,action_bar,
+              error_indicator,error_text}
             .init()
```
move this init to prev line, pls

wdanilo commented on src/rust/ide/view/graph-editor/src/component/node.rs:
```diff
@@ -366,9 +421,14 @@ impl NodeModel {
         self.drag_area.mod_position(|t| t.x = width/2.0);
 
+        self.error_indicator.set_position_x(width/2.0);
+        self.error_text.set_position_y(1.0 * height + TEXT_SIZE);
```
what is this `1.0 *`?

wdanilo commented on src/rust/ide/view/graph-editor/src/component/node.rs:
```diff
@@ -380,6 +440,18 @@ impl NodeModel {
         &self.visualization
     }
+
+    fn set_error(&self, error:&Option<Error>) {
```
don't use `&Option<Error>`, use `Option<&Error>` instead. This is a much better design, which allows for a much better usage. Imagine that in one point in your program you have `&Error`. Then using the second API you can bpass it here without cloning. The same applies to situation where you have `&Option<Error>` - you can pass it to second api without perf overhead, while the API you proposed is mmuch more limited.

wdanilo commented on src/rust/ide/view/graph-editor/src/component/node.rs:
```diff
@@ -380,6 +440,18 @@ impl NodeModel {
     }
+
+    fn set_error(&self, error:&Option<Error>) {
+        // TODO: This extra text field should not be required after #1026 has been finished.
+        //Instead we should get the error content as normal node output that is visible in the
+        //visualisation.
```
I think this comment shouldgo to the text field definition place. Moreover, I'm not sure this is true, as we want to display errors even with visualisations closed.

wdanilo replied:
alignment

wdanilo commented on src/rust/ide/view/graph-editor/src/component/node.rs:
```diff
@@ -391,6 +463,7 @@ impl Node {
         let selection = Animation::<f32>::new(network);
 
         let bg_color_anim = color::Animation::new(network);
+        let error_anim    = color::Animation::new(network);
```
these names are inconsistent now

wdanilo commented on src/rust/ide/view/graph-editor/src/component/node.rs:
```diff
@@ -180,6 +220,7 @@ ensogl::define_endpoints! {
         set_disabled        (bool),
         set_input_connected (span_tree::Crumbs,Option<Type>,bool),
         set_expression      (Expression),
+        set_error_status    (Option<Error>),
```
why not just `set_error`, like `set_expression` above, and the `set_error` function in model?

MichaelMauderer replied:
To indicate that this can also clear the error status by setting `None`. Just `set_error` does not imply that.

wdanilo replied:
call note: `set_error`

wdanilo commented on src/rust/ide/view/graph-editor/src/component/node.rs:
```diff
@@ -380,6 +440,18 @@ impl NodeModel {
+        //Instead we should get the error content as normal node output that is visible in the
+        //visualisation.
+        if let Some(error) = error.as_ref() {
+            self.error_text.set_content.emit(error.message.clone());
+        } else {
+            self.error_text.set_content.emit("".to_string());
```
refactor commmon part out, pls.

wdanilo commented on src/rust/ide/view/graph-editor/src/component/node.rs:
```diff
@@ -450,6 +523,28 @@ impl Node {
+                        let red:color::Lcha = color::Rgba::red().into();
+                        error_anim.target.emit(red);
+                    },
+                    None    => error_anim.target.emit(color::Lcha::transparent()),
```
alignment

wdanilo commented on src/rust/ide/view/graph-editor/src/component/node.rs:
```diff
@@ -450,6 +523,28 @@ impl Node {
+                match error.as_ref() {
+                    Some(error) => {
+                        let error_data:visualization::Data = error.clone().into();
+                        model.visualization.frp.set_data.emit(error_data);
+
+                        let red:color::Lcha = color::Rgba::red().into();
+                        error_anim.target.emit(red);
```
you are doing 2 frp emmits from these block, which means that this could be split into more fine-grained FRP endpoints. Please do it. Emitting more than one thing is almost always a bad design. Moreover, then you can not emit things at all, and use our new syntax here, like `error_anim.target <+ something`.

wdanilo commented on src/rust/ide/view/graph-editor/src/component/node.rs:
```diff
@@ -450,6 +523,28 @@ impl Node {
+            });
+
+            eval error_anim.value([model](value) {
+                let rgba = color::Rgba::from(value);
+                model.error_indicator.shape.color_rgba.set(rgba.into());
```
you can make it 1 line, then also remove curly braces, and also `[model]` part.

wdanilo commented on src/rust/ide/view/graph-editor/src/component/node.rs:
```diff
@@ -464,6 +559,7 @@ impl Node {
             );
         }
 
+        frp.set_error_status.emit(None);
```
why is it needed?

MichaelMauderer replied:
If we don't have it we get a black error indicator from the colour animation.

wdanilo commented on src/rust/ide/view/graph-editor/src/lib.rs:
```diff
@@ -362,6 +362,8 @@ ensogl::define_endpoints! {
         /// Remove all nodes from the graph.
         collapse_selected_nodes(),
+        /// Indicate whether this node had an error or not.
+        set_node_error_status((NodeId,Option<node::error::Error>)),
```
don't use double parenshere, our macros will handle that for you

wdanilo commented on src/rust/ide/view/graph-editor/src/lib.rs:
```diff
@@ -2424,11 +2426,25 @@ fn new_graph_editor(app:&Application) -> GraphEditor {
+    // === Set Node Error ===
+    frp::extend! { network
+
+    eval inputs.set_node_error_status([model]((node_id, error)) {
+        if let Some(node) = model.nodes.get_cloned_ref(node_id) {
+            node.set_error_status.emit(error)
+        }
+    });
```
Try not using `emit` fro inside `eval`. Currently, our FRP primitives are so rich, that it is not needed almost ever. Just use `<= ... . map(... model.nodes.get_cloned_ref(node_id))` and then zip it with error, and then `node.set_error_status <+ ...`

MichaelMauderer replied:
I'm not sure how you mean. Nodes can't be passed through the FRP network as they don't have a default value, so we need to unwrap in normal rust code.

wdanilo reviewed (changes requested):
Good to see it implemented! Few comments:

1. I'm surprised by the number of places where the code was not cleaned stylistically – remember to do self-review, please. It was not happening with previous PRs.

2. The resulting image looks bad, the lines are very pixelated, this needs to be fixed, however, I do not know what it is caused by (the equations look good to me). Definitely, something is broken here. Such an effect normally means that we are "shortening" the distance between pixels. Are you sure that without repeat, the single stripe would be of the same width than with the repeat? Also, is single stripe rendered correctly? Please show me that in DM on Discord and lets chat.
![image](https://user-images.githubusercontent.com/1623053/102852088-ba950200-441d-11eb-8ef5-042999a6463a.png)

3. No dark-mode screenshots! Please make some.

4. Also, please make bigger screenshots with the whole GUI and few nodes visible, so we can see how it looks like with normal zoom and with other nodes. Also, please test uch wider stripes below nodes.

MichaelMauderer commented:
>     2. The resulting image looks bad, the lines are very pixelated, this needs to be fixed, however, I do not know what it is caused by (the equations look good to me). Definitely, something is broken here. Such an effect normally means that we are "shortening" the distance between pixels. Are you sure that without repeat, the single stripe would be of the same width than with the repeat? Also, is single stripe rendered correctly? Please show me that in DM on Discord and lets chat.
>        ![image](https://user-images.githubusercontent.com/1623053/102852088-ba950200-441d-11eb-8ef5-042999a6463a.png)

This is most likely just the screen grab being overblown in the view here.
> 
>     3. No dark-mode screenshots! Please make some.

Added!

>     4. Also, please make bigger screenshots with the whole GUI and few nodes visible, so we can see how it looks like with normal zoom and with other nodes.
Added!

wdanilo commented on src/rust/ensogl/lib/core/src/display/shape/primitive/glsl/shape.glsl:
```diff
@@ -290,6 +296,14 @@ Shape set_color(Shape shape, Srgba t) {
+    Id       id    = s.id;
+    Color    color = s.color;
+    BoundSdf sdf   = s.sdf;
+    sdf.bounds = infinite();
```
alignment

wdanilo commented on src/rust/ensogl/lib/core/src/display/shape/primitive/glsl/shape.glsl:
```diff
@@ -322,3 +336,7 @@ vec2 scale (vec2 position, float value) {
   return vec2(length(position), atan(position.y, position.x));
 }
+
+vec2 repeat (vec2 position, vec2 t) {
+    return mod(position, t);
+}
```
"t" is not thebest name either ;)

wdanilo commented on src/rust/ensogl/lib/core/src/display/shape/primitive/shader/canvas.rs:
```diff
@@ -298,6 +298,20 @@ impl Canvas {
+
+    /// Repeat the shape with the given tile size.
+    pub fn repeat<T:Into<Var<Vector2<Pixels>>>>
+    (&mut self,  num:usize, s:Shape, tile_size:T) -> Shape {
```
double space

wdanilo commented on src/rust/ide/view/graph-editor/src/component/node.rs:
```diff
@@ -380,6 +450,16 @@ impl NodeModel {
+            error.message.clone()
+        } else {
+            "".to_string()
+        };
```
```
error.map(|t| t.message.clone()).unwrap_or_default()
```

wdanilo commented on src/rust/ide/view/graph-editor/src/component/node.rs:
```diff
@@ -450,6 +531,25 @@ impl Node {
+
+            eval error_color_anim.value((value) {
+                model.error_indicator.shape.color_rgba.set(color::Rgba::from(value).into());
+            });
```
rm curly braces

wdanilo commented on src/rust/ide/view/graph-editor/src/component/node/error.rs:
```diff
@@ -0,0 +1,34 @@
+///
+/// Note:
+/// Tis is a dummy implementation that can and should be extended once we have the proper
+/// error information from the language server. See #1026 for more information.
```
lets use full links
